### ISSUE TITLE
Some data structure (topology metadata related?) consumes a lot of memory (grows quickly as the cluster grows)

### ISSUE THREAD

**Issue by kbr-scylla** [scylladb/scylladb#19420]
In issue https://github.com/scylladb/scylladb/issues/18669

@aleksbykov ran a dtest that would boot 100 nodes, with mostly default settings for topology (256 vnodes per node).

The dtest would pass these smp/memory settings to Scylla:
```
    smp, (positional) 2, memory, (positional) 1024M,
```
which means we have 512M of memory per shard.

After booting around 70 nodes, nodes start running out of memory:
```
ERROR 2024-05-14 09:35:10,267 [shard 0: gms] lsa - Aborting due to allocation failure: failed to refill emergency reserve of 30 (have 12 free segments)
Aborting on shard 0.
Backtrace:
...
```

decoded relevant part:
```
logalloc::allocating_section::reserve(logalloc::tracker::impl&) at ./utils/logalloc.cc:2945
decltype(auto) logalloc::allocating_section::with_reserve<logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&)::{lambda()#1}>(logalloc::region, logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&)::{lambda()#1}) at ././utils/logalloc.hh:473
 (inlined by) decltype(auto) logalloc::allocating_section::operator()<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}>(logalloc::region&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0::operator()() const::{lambda()#1}&&) at ././utils/logalloc.hh:529
 (inlined by) operator() at ./replica/memtable.cc:800
 (inlined by) decltype(auto) with_allocator<replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0>(allocation_strategy&, replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&)::$_0&&) at ././utils/allocation_strategy.hh:318
 (inlined by) replica::memtable::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const> const&, db::rp_handle&&) at ./replica/memtable.cc:799
void replica::table::do_apply<frozen_mutation const&, seastar::lw_shared_ptr<schema const>&>(replica::compaction_group&, db::rp_handle&&, frozen_mutation const&, seastar::lw_shared_ptr<schema const>&) at ./replica/table.cc:2816
 (inlined by) operator() at ./replica/table.cc:2844
 (inlined by) seastar::future<void> seastar::futurize<void>::invoke<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&) at ././seastar/include/seastar/core/future.hh:2032
 (inlined by) auto seastar::futurize_invoke<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&) at ././seastar/include/seastar/core/future.hh:2066
 (inlined by) seastar::futurize<std::result_of<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0 ()>::type>::type replica::dirty_memory_manager_logalloc::region_group::run_when_memory_available<replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0>(replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >)::$_0&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/dirty_memory_manager.hh:572
 (inlined by) replica::table::apply(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/table.cc:2843
replica::database::apply_in_memory(frozen_mutation const&, seastar::lw_shared_ptr<schema const>, db::rp_handle&&, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >) at ./replica/database.cc:1833
replica::database::do_apply(seastar::lw_shared_ptr<schema const>, frozen_mutation const&, tracing::trace_state_ptr, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, seastar::bool_class<db::force_sync_tag>, std::variant<std::monostate, db::per_partition_rate_limit::account_only, db::per_partition_rate_limit::account_and_enforce>) at ./replica/database.cc:2053
```

so apparently some memtable grows large.

There's nothing else happening in this test apart from booting a large number of nodes.

So it seems that some topology-related metadata takes about 0.5G of memory after booting 70 nodes? Which sounds like a lot, and we should probably investigate.

This apparently happens both in 5.4 and 6.0, so it's not a (recent) regression.

I suggested using `1536M` instead of `1024M` and increase the instance size used for the test so that there's enough memory for all nodes with this new setting. Then the test managed to boot ~90 nodes before reaching the same error. From these two data points it seems the scaling is not linear, but hard to say anything with just two data points.

---

The goal of this issue is to investigate what actually consumes this memory in Scylla. If these are memtables, like the backtrace suggests -- why are they not flushed? But it could also be that we're consuming a lot of memory outside LSA regions which leaves little for memtables. Then we should research if this can be optimized somehow.

**Comment by michoecho** [comment]
Based on what I'm seeing in the logs, I claim with 70% confidence that the main memory consumer is `system.cdc_generations_v3`.

**Comment by kbr-scylla** [comment]
This also reproduced in 5.4, which doesn't have `system.cdc_generations_v3`.

**Comment by michoecho** [comment]
> This also reproduced in 5.4, which doesn't have `system.cdc_generations_v3`.

It has `system.cdc_generations_v2`, though.

**Comment by kbr-scylla** [comment]
> It has system.cdc_generations_v2, though.

If you meant `system_distributed_everywhere.cdc_generation_descriptions_v2`, then yes.

But the memtables should be flushed under memory pressure, and we should reach hard limit if it's not flushed in time that would prevent further writes from happening. So looks like we need to investigate why memtable is not being flushed and why we don't block on writing to the table before that happens

**Comment by michoecho** [comment]
> > It has system.cdc_generations_v2, though.
> 
> If you meant `system_distributed_everywhere.cdc_generation_descriptions_v2`, then yes.
> 
> But the memtables should be flushed under memory pressure, and we should reach hard limit if it's not flushed in time that would prevent further writes from happening. So looks like we need to investigate why memtable is not being flushed and why we don't block on writing to the table before that happens

Bold of you to assume that there is only one problem here. (And even bolder to assume that it has something to do with memtables not being flushed).

After looking at some cores, I see at least two independent OOM causes:
- One is about the LSA reserve getting unreasonably bloated when `cdc_generation_descriptions_v3` memtables are merged into the cache. It's weird, because the size of reserve is bound by the amount of memory allocated during a single task quota, and ~256 MiB should be too much to be allocated within a millisecond. Maybe something unintended is happening there. 
- One is about the quadratic behaviour of raft replication. AFAIU when a new raft entry appears, the leader sends it to all nodes, which involves creating N (number of nodes) copies of the message in RAM. So the memory usage is N * message size. And since the CDC generation message size grows up to a few megabytes, and N=100, this requires a few hundred megabytes.

**Comment by kbr-scylla** [comment]
> One is about the quadratic behaviour of raft replication. AFAIU when a new raft entry appears, the leader sends it to all nodes, which involves creating N (number of nodes) copies of the message in RAM. So the memory usage is N * message size. And since the CDC generation message size grows up to a few megabytes, and N=100, this requires a few hundred megabytes.

- this does not apply to 5.4
- in 6.0: we don't send the entire cdc generation in a single message, we split into parts based on raft command size limit, which is calculated from commitlog segment size limit (we can split further if needed)

**Comment by kbr-scylla** [comment]
But I guess similar argument could apply to the storage_proxy system_distributed_everywhere write we do in 5.4

**Comment by kbr-scylla** [comment]
The `system_distributed_everywhere` logic used by 5.4 is written to limit memory consumption (`system_distributed_keyspace::insert_cdc_generation`)

**Comment by kbr-scylla** [comment]
All of this makes little sense

Size of CDC generation for 100 node cluster with smp=2 is:
`256 * 100 * 2 * 16 = 819200` B = 800 KB

Even if we kept ALL 100 generations (one created per node joining) in memory (which we don't, we maybe keep last 2) it would be `256 * (100 * 99 / 2) * 2 * 16` ~= 38 MB

Mutation metadata giving over 10x overhead? Unlikely

Could it be that we're not actually consuming .5G of memory -- maybe this allocation failure is actually related to the system memory pool for memtables, which is 10MB, the failure happens when we try to exceed that 10MB

which means the problem is not that big and we could probably solve it lazily by increasing the pool size to 20MB

But still, I don't get one thing, when we try to apply to memtable and there's not enough memory, it shouldn't cause failure, it should block on a semaphore, which is then unblocked by the "dirty memory manager" when memtables are flushed

---

@michoecho when you were looking at cores, did you actually verify the .5G that I mentioned in the original post?

**Comment by michoecho** [comment]
> @michoecho when you were looking at cores, did you actually verify the .5G that I mentioned in the original post?

I don't understand the question.

Besides, this issue has already been evicted from my memory.

**Comment by kbr-scylla** [comment]
Did you see that something consumes nearly 500MB of memory

also you never said why you suspect `system.cdc_generations_v3`

**Comment by michoecho** [comment]
> also you never said why you suspect `system.cdc_generations_v3`

I did, in https://github.com/scylladb/scylladb/issues/19420#issuecomment-2186538039.

In one core I observed the memory being filled with Raft messages with CDC generations, and in a different core the memtable flush which exploded the LSA reserve was also a cdc_generations memtable.

> did you actually verify the .5G
> Did you see that something consumes nearly 500MB of memory

What kind of question is that? If the node failed with an OOM, then something consumed the entire memory. Do you suspect that the allocator is broken?

Also, as I said, there are at least two separate problems here, and this ticket conflates them.

One issue, as far as I can see, is about the quadratic behaviour of raft, and in this case the memory is indeed exhausted.

In the other issue, where the reserve explodes to ~1024 segments, the memory isn't exhausted — something weird is happening there, the reserve isn't supposed to grow so much.

> Could it be that we're not actually consuming .5G of memory -- maybe this allocation failure is actually related to the system memory pool for memtables, which is 10MB, the failure happens when we try to exceed that 10MB

I'm not sure where you're taking that 10 MiB figure from, but there is no such thing as a "memory pool for memtables". Memtables share their memory pool with the rest of the node, there is no "physical" separation between the memtable allocator and other allocators. If you see `failed to refill emergency reserve of 30 (have 12 free segments)`, it means that the node as a whole is out of memory.

**Comment by kbr-scylla** [comment]
>  If you see failed to refill emergency reserve of 30 (have 12 free segments), it means that the node as a whole is out of memory.

OK

**Comment by kbr-scylla** [comment]
> I'm not sure where you're taking that 10 MiB figure from

```cpp
    , _system_dirty_memory_manager(*this, 10 << 20, cfg.unspooled_dirty_soft_limit(), default_scheduling_group())
    , _dirty_memory_manager(*this, dbcfg.available_memory * 0.50, cfg.unspooled_dirty_soft_limit(), dbcfg.statement_scheduling_group)
```

`10 << 20`

(`database::database` constructor)

**Comment by michoecho** [comment]
> > I'm not sure where you're taking that 10 MiB figure from
> 
> ```c++
>     , _system_dirty_memory_manager(*this, 10 << 20, cfg.unspooled_dirty_soft_limit(), default_scheduling_group())
>     , _dirty_memory_manager(*this, dbcfg.available_memory * 0.50, cfg.unspooled_dirty_soft_limit(), dbcfg.statement_scheduling_group)
> ```
> 
> `10 << 20`
> 
> (`database::database` constructor)

Oh, ok. I didn't even realize that system tables have a separate controller.

But, as I said, the dirty memory manager is on a layer above the allocators, and it's a "soft" thing. It controls how fast memtables are flushed and whether we should block further writes from going into the memtables or not, but it doesn't prevent any allocations from happening, so it shouldn't be relevant in an OOM issue.

**Comment by avikivity** [comment]
Could it be that compaction isn't able to keep up, and we need to read from 100 sstables? Though even that shouldn't be too bad.

**Comment by gusev-p** [comment]
The same problem reproduces without raft topology -- I ran the [same test job with --force-gossip-topology-changes](https://jenkins.scylladb.com/job/scylla-master/job/byo/job/byo_build_tests_dtest/2759), got the same crash on 34-th node:
```
INFO  2025-02-18 14:32:18,610 [shard 1:strm] repair - repair[aec17093-c2aa-413c-8778-390d5fe61510]: Started to repair 4 out of 4 tables in keyspace=system_distributed, table=cdc_streams_descriptions_v2, table_id=0bf73fd7-65b2-36b0-85e5-658131d5df36, repair_reason=bootstrap
ERROR 2025-02-18 14:32:19,341 [shard 0:sl:d] lsa - Aborting due to allocation failure: failed to refill emergency reserve of 30 (have 28 free segments)
Aborting on shard 0, in scheduling group sl:default.
```

**Comment by kostja** [comment]
@avikivity suggests that we need to move all system_distributed keysapces to use tablets. This will speed up streaming.

**Comment by mykaul** [comment]
@aleksbykov - can we see if this is still an issue now with master (with CDCv3 and everything) ?

**Comment by aleksbykov** [comment]
@mykaul ,

Something similar happened on latest masters: https://jenkins.scylladb.com/job/sct-github-PRs-scan/job/scylla-dtest/job/PR-5666/54/, and mostly it depend on runner instance size.

i will take some large instance on aws and will run similar test with 100 nodes.

**Comment by mykaul** [comment]
@aleksbykov - I think latest tests - we are good, no?
